### PR TITLE
Stop using deprecated cuda.empty_like and cuda.empty

### PR DESCRIPTION
**Pull request by wkentaro**

### CODE REVIEW

**Comment by wkentaro** [comment]
related to https://github.com/pfnet/chainer/issues/707 ?
```

Successfully installed flake8-2.5.1 hacking-0.10.2 mccabe-0.3.1 pbr-1.8.1 pep8-1.5.7 pyflakes-1.0.0
travis_time:end:08d15558:start=1449951990555248841,finish=1449951991777097894,duration=1221849053
[0Ktravis_fold:end:install.2
[0Ktravis_time:start:007ad244
[0K$ nosetests -a '!gpu' tests/chainer_tests
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
======================================================================
ERROR: Failure: ImportError (cannot import name core)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/nose/loader.py", line 414, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/travis/build/pfnet/chainer/tests/chainer_tests/functions_tests/activation_tests/test_clipped_relu.py", line 5, in <module>
    import chainer
  File "/home/travis/build/pfnet/chainer/chainer/__init__.py", line 6, in <module>
    from chainer.functions import basic_math
  File "/home/travis/build/pfnet/chainer/chainer/functions/__init__.py", line 7, in <module>
    from chainer.functions.activation import relu
  File "/home/travis/build/pfnet/chainer/chainer/functions/activation/relu.py", line 7, in <module>
    import cupy
  File "/home/travis/build/pfnet/chainer/cupy/__init__.py", line 8, in <module>
    from cupy import binary
  File "/home/travis/build/pfnet/chainer/cupy/binary/__init__.py", line 5, in <module>
    from cupy.binary import elementwise  # NOQA
  File "/home/travis/build/pfnet/chainer/cupy/binary/elementwise.py", line 1, in <module>
    from cupy import core
  File "/home/travis/build/pfnet/chainer/cupy/core/__init__.py", line 1, in <module>
    from cupy.core import core
ImportError: cannot import name core

```

**Comment by wkentaro** [comment]
Now I think it's fixed. Needed below line.
https://github.com/pfnet/chainer/pull/755/files#diff-7a6eaebedbbf7319fab629d0d85c5b41R8

**Comment by unnonouno** [comment]
Thank you very much! As you mentioned, these methods are deprecated.
Please use `cuda.cupy.empty_like` instead of `cupy.empty_like`. `if cuda.available` is not necessary.

**Comment by wkentaro** [comment]
You mean this can be removed?
https://github.com/pfnet/chainer/pull/755/files#diff-7a6eaebedbbf7319fab629d0d85c5b41R8

but I had error at nosetests on travis.

**Comment by unnonouno** [comment]
`cuda.cupy` is equivalent to `cupy` only when `cuda.available` is `True`. When you use `cupy` directly, it causes an error on an environment without CUDA.

**Comment by wkentaro** [comment]
Thanks, I think I understood.
I updated the commit, please review.

File: chainer/functions/math/matmul.py [hunk 1]
```diff
@@ -244,8 +244,8 @@ def forward_gpu(self, x):
     def backward_gpu(self, x, gy):
         a, b = x
         batch_size = a.shape[0]
-        ga = cuda.empty((batch_size,) + _as_mat(a[0]).shape)
-        gb = cuda.empty((batch_size,) + _as_mat(b[0]).shape)
+        ga = cuda.cupy.empty((batch_size,) + _as_mat(a[0]).shape)
+        gb = cuda.cupy.empty((batch_size,) + _as_mat(b[0]).shape)
```
Comment (unnonouno) [inline chainer/functions/math/matmul.py]: Please set `dtype`. I mean `ga = cuda.cupy.empty(..., a.dtype)`.
Default `dtype` of `cupy.empty` is `float`, is not `float32`.

**Comment by unnonouno** [comment]
Yes, great!

Please check your patch 
- run `nosetests` and find errors
  - http://docs.chainer.org/en/stable/contribution.html#testing-guidelines
  - if you only have one GPU, run `nosetests   -A 'gpu<2'`
- check code style with `flake8` and `hacking`
  - http://docs.chainer.org/en/stable/contribution.html#coding-guidelines
  - `pip install flake8 hacking && flake8 chainer`

**Comment by unnonouno** [comment]
Thank you for quick fix! It looks good.

**Pull request merged**